server/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java

```diff
@@ -158,21 +159,33 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
         }
         // If we're using the default keyword analyzer then we can rewrite this to a TermQueryBuilder
         // and possibly shortcut
-        if (analyzer != null) {
-            if (sec.getIndexAnalyzers().get(analyzer) == Lucene.KEYWORD_ANALYZER) {
-                TermQueryBuilder termQueryBuilder = new TermQueryBuilder(fieldName, value);
-                return termQueryBuilder.rewrite(queryRewriteContext);
-            }
-        } else {
-            MappedFieldType mft = sec.getFieldType(fieldName);
-            if (mft != null && mft.getTextSearchInfo().getSearchAnalyzer() == Lucene.KEYWORD_ANALYZER) {
-                TermQueryBuilder termQueryBuilder = new TermQueryBuilder(fieldName, value);
-                return termQueryBuilder.rewrite(queryRewriteContext);
+        // If we're using a keyword analyzer then we can rewrite this to a TermQueryBuilder
+        // and possibly shortcut
+        NamedAnalyzer configuredAnalyzer = configuredAnalyzer(sec);
+        if (configuredAnalyzer != null && configuredAnalyzer.analyzer() instanceof KeywordAnalyzer) {
+            if (value.toString().isEmpty()) {
+                if (zeroTermsQuery == ZeroTermsQueryOption.ALL) {
+                    return new MatchAllQueryBuilder();
+                }
+                return new MatchNoneQueryBuilder();
```

Comment:
Can we use zeroTermsQuery#asQuery() instead if the if statement here? That would include the NULL option which returns null, I don't know if this can cause problems in further rewriting.